### DISCUSSION TITLE
Stable short link to "editable" code?

### DISCUSSION THREAD

**Discussion by jll63** [compiler-explorer/compiler-explorer#6352] [Q&A]
I would like to embed links to CE examples of my YOMM2 library in my GitHub pages, in social media posts, and in presentations. I would also like to be able to modify the examples (say, to correct something or improve a comment), and have the embedded link point to the new version. Unless I misunderstand, short links will not give me this: whenever I click "share" after a change, a new short link is generated.

I guess I could create HTML redirection pages  to forward to CE links, and update them as needed. But surely, I am not the first to want to do this. Does anybody have an easier/better solution?

**Comment by mattgodbolt** [comment]
Shared links are immutable (and heavily cached); so yes, you need to make a new short link every time.

There's two possibilities:
- Use a full link (which you can encode yourself, with some work)
- Use our shortener API to create short links yourself

I know @lefticus does the latter; to have links that on `make` or whatever in a presentation auto-update (though we ask that you cache locally to avoid hitting our API for unchanged code).  Some docs on the APIs at https://github.com/compiler-explorer/compiler-explorer/blob/main/docs/API.md#post-apishortener---saves-given-state-forever-to-a-shortlink-and-returns-the-unique-id-for-the-link

**Reply by lefticus** [reply]
Correct, I just dynamically build mine and I do try my best to cache everything in the build process. My solution does not work for those links that might get baked into books.

**Reply by mattgodbolt** [reply]
We have a long standing "request" for reassignable links, which is definitely filed under "One Day™", specifically for things we'd like to otherwise change that (e.g.) are in books. It's a manual process on my end doing that right now (though the underlying system sort of supports it)

**Comment by partouf** [comment]
If you have short enough blocks of code, you could consider producing `/clientstate/<base64>` links https://github.com/compiler-explorer/compiler-explorer/blob/main/docs/API.md#get-clientstatebase64---opens-the-website-in-a-given-state

Of course they will be long links and not very user friendly.

But you can have some JS magic in case of website or in case of github docs you can make a script that can read the code blocks and combine that with a clientstate template and add a link.

The base64 in the link is the same data as you would send with `/api/shortener`, but you wouldn't have to `POST` anything to create the link.

**Reply by partouf** [reply]
Foonathan wrote a blog about this back in 2021 that demonstrates this https://www.foonathan.net/2021/05/hugo-godbolt/

**Comment by jll63** [comment]
Thanks everybody for the suggestions.

In my case the URLs need to be reasonably short (they have to fit on a slide), and be (or seem) permanent. Also, I want to be able to generate QR codes for them, so the length matters.

For the time being, I am using HTML redirection pages, like https://jll63.github.io/yomm2/ce/ast.html, which contains [this](https://github.com/jll63/yomm2/blob/master/docs/ce/ast.html). They don't have the cachet of explicit godbolt.org though.
